opensearch_single_kernel/managers/lock.py

```diff
@@ -0,0 +1,507 @@
+
+                    # This unit has OpenSearch lock
+                    unit = self.state.opensearch_unit.unit_name
+
+            if unit == self.state.opensearch_unit.unit_name:
+                # Lock acquired
+                # Release peer databag lock, if any
+                self.logger.debug("[Node lock] Acquired via opensearch")
+                super().release()
+                self.logger.debug("[Node lock] Released redundant peer lock (if held)")
+                return True
+
+            if unit:
+                # Another unit has lock
+                self.logger.debug(f"[Node lock] Not acquired. Unit with opensearch lock: {unit}")
+                return False
+
+            assert online_nodes == 1
+            self.logger.debug("[Node lock] No unit has opensearch lock")
+        self.logger.debug("[Node lock] Using peer databag for lock")
+        # Request peer databag lock
+        # If return value is True:
+        # - Lock granted in previous Juju event
+        # - OR, unit is leader & lock granted in this Juju event
+        return super().acquired
+
+    def release(self):
+        """Release lock.
+
+        Limitation: if lock acquired via OpenSearch document and all units offline, OpenSearch
+        document lock will not be released
+        """
+        self.logger.debug("[Node lock] Releasing lock")
+
+        # fetch current app description
+        # TODO: Add peer CM
+        current_app = self._charm.opensearch_peer_cm.deployment_desc().app
+
+        host = self.state.host_ip if self.opensearch_client.is_node_up() else None
+        alt_hosts = self.state.alt_hosts
+        if host or alt_hosts:
+            self.logger.debug("[Node lock] Checking which unit has opensearch lock")
+            # Check if this unit currently has lock
+            # or if there is a stale lock from a unit no longer existing
+            # for large deployments the MAIN/FAILOVER orchestrators should broadcast info
+            # over non-online units in the relation. This info should be considered here as well.
+            unit_with_lock = self._unit_with_lock(host)
+            current_app_units = [
+                format_unit_name(unit, app=current_app) for unit in self.state.all_units
+            ]
+
+            # TODO:
+            # handle case of large deployments
+            other_apps_units = []
+            # if all_apps := self._charm.peers_data.get_object(Scope.APP, "cluster_fleet_apps"):
+            # for app in all_apps.values():
+            # p_cluster_app = PeerClusterApp.from_dict(app)
+            # if p_cluster_app.app.id == current_app.id:
+            # continue
+
+            # units = [
+            # format_unit_name(unit, app=p_cluster_app.app)
+            # for unit in p_cluster_app.units
+            # ]
+            # other_apps_units.extend(units)
+
+            if unit_with_lock and (
+                unit_with_lock == self.state.opensearch_unit.unit_name
```

Comment:
```suggestion
              unit_with_lock == self.state.unit_name
```
 self.state.opensearch_unit.unit_name is not available.